classes/Tools.php

```diff
@@ -792,10 +792,8 @@ protected static function getContextLocale(Context $context)
             return $locale;
         }
 
-        $container = isset($context->controller) ? $context->controller->getContainer() : null;
-        if (null === $container) {
-            $container = SymfonyContainer::getInstance();
-        }
+        $containerFinder = new ContainerFinder();
```

Comment:
Hi, I don't understand the need for this `ContainerFinder` it seems to be hiding some magic about how to get the container, and can return:
- the container from the Context property (which isn't even defined)
- container from the controller (if you are luck to have one in the context)
- the Symfony container as a fallback

I know this will work but in the end you don't even know which container you will have.
I would suggest simply using the `ContainerBuilder` as you use it (which is the right approach I think)

```php
$container = ContainerBuilder::getContainer('webservice', _PS_MODE_DEV_);
```

This way you are sure to have the container defined for the webservice environment, and don't worry it won't be instanciated several times because `ContainerBuilder` stores the instanciated containers by environment name

Reply:
Oh my bad I didn't notice all this is made in the Tools class..

Reply:
@jolelievre But in Tools, you don't know of you need webservice/front/admin container.

Reply:
Yep I realized it a bit later ^^

Reply:
@jolelievre Finally Changes or Approved ? :)

Reply:
Not a big fan that we need to do this.. but I don't see another solution for now So I'll approve
I believe @eternoendless has a few feedback not answered yet

Reply:
I understand. But after discussion with @mickaelandrieu, it's the cleanest solution to offer.